File: java/src/main/java/io/github/amisonnet8/sandbox/internal/codec/Decode.java
package io.github.amisonnet8.sandbox.internal.codec;

import io.github.amisonnet8.sandbox.DumpResult;
import io.github.amisonnet8.sandbox.ExecResult;
import io.github.amisonnet8.sandbox.Hello;
import io.github.amisonnet8.sandbox.InspectResult;
import io.github.amisonnet8.sandbox.QueryResult;
import io.github.amisonnet8.sandbox.SchemaResult;
import io.github.amisonnet8.sandbox.SnapshotResult;
import io.github.amisonnet8.sandbox.TablesResult;

import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;

/**
 * Decoding: turning wire JSON (parsed into a {@link Json} tree by {@link
 * JsonParser}) into this library's public result records, defined one
 * package up ({@code io.github.amisonnet8.sandbox}) since they are part of
 * the supported API and this package is not.
 */
public final class Decode {

    private Decode() {
    }

    /** Decodes a hello line. Does not validate the {@code protocol} number: that is the caller's job. */
    public static Hello hello(byte[] line) {
        Json v = parseLine(line, "hello");
        requireObject(v, "hello line");
        return new Hello((int) requiredInt(v, "protocol"), requiredStr(v, "version"), requiredStr(v, "product"));
    }

    /** Decodes one response line into (ok, the full field set). */
    public static DecodedResponse responseLine(byte[] line) {
        Json v = parseLine(line, "response");
        requireObject(v, "response line");
        Boolean ok = asBool(get(v, "ok"));
        return new DecodedResponse(ok != null && ok, v);
    }

    /** Extracts {@code {"error":{"code":...,"message":...}}} from a response, or {@code null} if absent/malformed. */
    public static ResponseError error(DecodedResponse r) {
        Json err = get(r.fields, "error");
        if (err == null) {
            return null;
        }
        String code = asStr(get(err, "code"));
        String message = asStr(get(err, "message"));
        if (code == null || message == null) {
            return null;
        }
        return new ResponseError(code, message);
    }

    public static QueryResult query(DecodedResponse r) {
        return new QueryResult(requiredStrList(r.fields, "columns"), decodeRows(required(r.fields, "rows")));
    }

    public static ExecResult exec(DecodedResponse r) {
        return new ExecResult(requiredInt(r.fields, "rows_affected"), requiredInt(r.fields, "last_insert_id"));
    }

    public static SnapshotResult snapshot(DecodedResponse r) {
        return new SnapshotResult(requiredStr(r.fields, "path"));
    }

    public static InspectResult inspect(DecodedResponse r) {
        return new InspectResult(
                requiredBool(r.fields, "has_data"),
                requiredOptionalLong(r.fields, "version"),
                requiredOptionalLong(r.fields, "data_length"),
                requiredStr(r.fields, "source"),
                requiredBool(r.fields, "read_only"));
    }

    public static TablesResult tables(DecodedResponse r) {
        return new TablesResult(requiredStrList(r.fields, "tables"));
    }

    public static SchemaResult schema(DecodedResponse r) {
        return new SchemaResult(requiredStrList(r.fields, "schema"));
    }

    public static DumpResult dump(DecodedResponse r) {
        return new DumpResult(requiredStr(r.fields, "sql"));
    }

    // --- value/rows decoding ---

    static Object decodeValue(Json raw) {
        if (raw instanceof Json.Null) {
            return null;
        }
        if (raw instanceof Json.Bool b) {
            throw new CodecException("unexpected boolean value in response: " + b.value());
        }
        if (raw instanceof Json.Num n) {
            NumberToken token = n.token();
            if (token.isReal()) {
                return token.asDouble();
            }
            return token.asLong();
        }
        if (raw instanceof Json.Str s) {
            return s.value();
        }
        if (raw instanceof Json.Arr a) {
            List<Json> items = a.items();
            if (items.size() != 1) {
                throw new CodecException(
                        "protocol violation: BLOB array must have exactly 1 element, got " + items.size());
            }
            String s = asStr(items.get(0));
            if (s == null) {
                throw new CodecException("protocol violation: BLOB array element is not a string");
            }
            return B64.decode(s);
        }
        throw new CodecException("unexpected object value in response");
    }

    static List<List<Object>> decodeRows(Json raw) {
        if (!(raw instanceof Json.Arr arr)) {
            throw new CodecException("rows field is not an array");
        }
        List<List<Object>> rows = new ArrayList<>(arr.items().size());
        int i = 0;
        for (Json row : arr.items()) {
            if (!(row instanceof Json.Arr rowArr)) {
                throw new CodecException("rows[" + i + "] is not an array");
            }
            List<Object> decoded = new ArrayList<>(rowArr.items().size());
            for (Json cell : rowArr.items()) {
                try {
                    decoded.add(decodeValue(cell));
                } catch (CodecException e) {
                    throw new CodecException("rows[" + i + "]: " + e.getMessage(), e);
                }
            }
            rows.add(decoded);
            i++;
        }
        return rows;
    }

    // --- Json tree accessors ---

    private static Json parseLine(byte[] line, String what) {
        try {
            return JsonParser.parse(new String(line, StandardCharsets.UTF_8));
        } catch (CodecException e) {
            throw new CodecException("invalid " + what + " line: " + e.getMessage(), e);
        }
    }

    private static void requireObject(Json v, String what) {
        if (!(v instanceof Json.Obj)) {
            throw new CodecException(what + " is not a JSON object");
        }
    }

    static Json get(Json v, String key) {
        if (!(v instanceof Json.Obj o)) {
            return null;
        }
        for (Json.Obj.Member m : o.members()) {
            if (m.key().equals(key)) {
                return m.value();
            }
        }
        return null;
    }

    private static Boolean asBool(Json v) {
        return (v instanceof Json.Bool b) ? b.value() : null;
    }

    private static String asStr(Json v) {
        return (v instanceof Json.Str s) ? s.value() : null;
    }

    private static Json required(Json v, String key) {
        Json field = get(v, key);
        if (field == null) {
            throw new CodecException("response missing field \"" + key + "\"");
        }
        return field;
    }

    private static String requiredStr(Json v, String key) {
        String s = asStr(required(v, key));
        if (s == null) {
            throw new CodecException("response field \"" + key + "\" is not a string");
        }
        return s;
    }

    private static boolean requiredBool(Json v, String key) {
        Boolean b = asBool(required(v, key));
        if (b == null) {
            throw new CodecException("response field \"" + key + "\" is not a bool");
        }
        return b;
    }

    private static long requiredInt(Json v, String key) {
        Json field = required(v, key);
        if (!(field instanceof Json.Num n)) {
            throw new CodecException("response field \"" + key + "\" is not an integer");
        }
        return n.token().asLong();
    }

    private static Long requiredOptionalLong(Json v, String key) {
        Json field = required(v, key);
        if (field instanceof Json.Null) {
            return null;
        }
        if (!(field instanceof Json.Num n)) {
            throw new CodecException("response field \"" + key + "\" is not an integer or null");
        }
        return n.token().asLong();
    }

    private static List<String> requiredStrList(Json v, String key) {
        Json field = required(v, key);
        if (!(field instanceof Json.Arr arr)) {
            throw new CodecException("response field \"" + key + "\" is not an array");
        }
        List<String> out = new ArrayList<>(arr.items().size());
        for (Json item : arr.items()) {
            String s = asStr(item);
            if (s == null) {
                throw new CodecException("response field \"" + key + "\" is not a string array");
            }
            out.add(s);
        }
        return out;
    }
}

File: rust/src/codec/decode.rs
//! Decoding: turning wire JSON (already parsed into a [`Json`] tree by
//! `codec::json`) into this crate's `Value`/result types.

use crate::error::Error;
use crate::value::{Row, Value};

use super::base64;
use super::json::{self, Json};

/// The connection's own greeting, read once before any request is sent.
#[derive(Debug, Clone, PartialEq)]
pub struct Hello {
    pub protocol: i64,
    pub version: String,
    pub product: String,
}

#[derive(Debug, Clone, PartialEq)]
pub struct QueryResult {
    pub columns: Vec<String>,
    pub rows: Vec<Row>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct ExecResult {
    pub rows_affected: i64,
    pub last_insert_id: i64,
}

#[derive(Debug, Clone, PartialEq)]
pub struct SnapshotResult {
    pub path: String,
}

#[derive(Debug, Clone, PartialEq)]
pub struct InspectResult {
    pub has_data: bool,
    pub version: Option<i64>,
    pub data_length: Option<i64>,
    pub source: String,
    pub read_only: bool,
}

#[derive(Debug, Clone, PartialEq)]
pub struct TablesResult {
    pub tables: Vec<String>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct SchemaResult {
    pub schema: Vec<String>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct DumpResult {
    pub sql: String,
}

/// One decoded response line: whether it was `ok`, and the full parsed
/// object (every top-level field, including `"ok"` itself) -- op decoders
/// below only look at the fields relevant to them, but the conformance
/// runner's partial-match comparisons need the whole thing.
pub(crate) struct DecodedResponse {
    pub(crate) ok: bool,
    pub(crate) fields: Json,
}

/// Decodes a hello line. Does not validate the protocol number -- that's
/// the caller's job, since only it knows what happens next (protocol.md).
pub(crate) fn decode_hello(line: &[u8]) -> Result<Hello, Error> {
    let v = json::parse(line).map_err(|e| Error::Protocol(format!("invalid hello line: {e}")))?;
    if v.as_object().is_none() {
        return Err(Error::Protocol(
            "hello line is not a JSON object".to_string(),
        ));
    }
    Ok(Hello {
        protocol: required_int(&v, "protocol")?,
        version: required_str(&v, "version")?.to_string(),
        product: required_str(&v, "product")?.to_string(),
    })
}

/// Decodes one response line into (ok, the full field set). Callers use
/// [`response_error`] on `fields` when `ok` is false.
pub(crate) fn decode_response_line(line: &[u8]) -> Result<DecodedResponse, Error> {
    let v =
        json::parse(line).map_err(|e| Error::Protocol(format!("invalid response line: {e}")))?;
    if v.as_object().is_none() {
        return Err(Error::Protocol(
            "response line is not a JSON object".to_string(),
        ));
    }
    let ok = v.get("ok").and_then(Json::as_bool).unwrap_or(false);
    Ok(DecodedResponse { ok, fields: v })
}

/// Extracts `{"error":{"code":...,"message":...}}` from a response's
/// fields, if present and well-formed.
pub(crate) fn response_error(fields: &Json) -> Option<Error> {
    let err = fields.get("error")?;
    let code = err.get("code")?.as_str()?;
    let message = err.get("message")?.as_str()?;
    Some(Error::Response {
        code: code.to_string(),
        message: message.to_string(),
    })
}

pub(crate) fn decode_value(raw: &Json) -> Result<Value, Error> {
    match raw {
        Json::Null => Ok(Value::Null),
        Json::Bool(b) => Err(Error::Protocol(format!(
            "unexpected boolean value in response: {b}"
        ))),
        Json::Number(n) => {
            if n.is_real() {
                let f = n.as_f64().ok_or_else(|| {
                    Error::Protocol(format!("invalid REAL token in response: {}", n.as_str()))
                })?;
                Ok(Value::Real(f))
            } else {
                let i = n.as_i64().ok_or_else(|| {
                    Error::Protocol(format!(
                        "integer value out of int64 range in response: {}",
                        n.as_str()
                    ))
                })?;
                Ok(Value::Integer(i))
            }
        }
        Json::String(s) => Ok(Value::Text(s.clone())),
        Json::Array(items) => {
            if items.len() != 1 {
                return Err(Error::Protocol(format!(
                    "protocol violation: BLOB array must have exactly 1 element, got {}",
                    items.len()
                )));
            }
            let s = items[0].as_str().ok_or_else(|| {
                Error::Protocol(
                    "protocol violation: BLOB array element is not a string".to_string(),
                )
            })?;
            let bytes = base64::decode(s)
                .map_err(|e| Error::Protocol(format!("invalid BLOB base64: {e}")))?;
            Ok(Value::Blob(bytes))
        }
        Json::Object(_) => Err(Error::Protocol(
            "unexpected object value in response".to_string(),
        )),
    }
}

pub(crate) fn decode_rows(raw: &Json) -> Result<Vec<Row>, Error> {
    let arr = raw
        .as_array()
        .ok_or_else(|| Error::Protocol("rows field is not an array".to_string()))?;
    let mut rows = Vec::with_capacity(arr.len());
    for (i, row) in arr.iter().enumerate() {
        let row_arr = row
            .as_array()
            .ok_or_else(|| Error::Protocol(format!("rows[{i}] is not an array")))?;
        let mut decoded = Vec::with_capacity(row_arr.len());
        for v in row_arr {
            decoded.push(decode_value(v).map_err(|e| match e {
                Error::Protocol(m) => Error::Protocol(format!("rows[{i}]: {m}")),
                other => other,
            })?);
        }
        rows.push(decoded);
    }
    Ok(rows)
}

fn required<'a>(v: &'a Json, key: &str) -> Result<&'a Json, Error> {
    v.get(key)
        .ok_or_else(|| Error::Protocol(format!("response missing field {key:?}")))
}

fn required_str<'a>(v: &'a Json, key: &str) -> Result<&'a str, Error> {
    required(v, key)?
        .as_str()
        .ok_or_else(|| Error::Protocol(format!("response field {key:?} is not a string")))
}

fn required_bool(v: &Json, key: &str) -> Result<bool, Error> {
    required(v, key)?
        .as_bool()
        .ok_or_else(|| Error::Protocol(format!("response field {key:?} is not a bool")))
}

fn required_int(v: &Json, key: &str) -> Result<i64, Error> {
    let n = required(v, key)?
        .as_number()
        .ok_or_else(|| Error::Protocol(format!("response field {key:?} is not an integer")))?;
    n.as_i64().ok_or_else(|| {
        Error::Protocol(format!(
            "response field {key:?} is not a valid int64: {}",
            n.as_str()
        ))
    })
}

fn required_optional_int(v: &Json, key: &str) -> Result<Option<i64>, Error> {
    let f = required(v, key)?;
    if f.is_null() {
        return Ok(None);
    }
    let n = f.as_number().ok_or_else(|| {
        Error::Protocol(format!("response field {key:?} is not an integer or null"))
    })?;
    Ok(Some(n.as_i64().ok_or_else(|| {
        Error::Protocol(format!(
            "response field {key:?} is not a valid int64: {}",
            n.as_str()
        ))
    })?))
}

fn required_str_list(v: &Json, key: &str) -> Result<Vec<String>, Error> {
    let arr = required(v, key)?
        .as_array()
        .ok_or_else(|| Error::Protocol(format!("response field {key:?} is not an array")))?;
    let mut out = Vec::with_capacity(arr.len());
    for item in arr {
        let s = item.as_str().ok_or_else(|| {
            Error::Protocol(format!("response field {key:?} is not a string array"))
        })?;
        out.push(s.to_string());
    }
    Ok(out)
}

pub(crate) fn decode_query_response(fields: &Json) -> Result<QueryResult, Error> {
    Ok(QueryResult {
        columns: required_str_list(fields, "columns")?,
        rows: decode_rows(required(fields, "rows")?)?,
    })
}

pub(crate) fn decode_exec_response(fields: &Json) -> Result<ExecResult, Error> {
    Ok(ExecResult {
        rows_affected: required_int(fields, "rows_affected")?,
        last_insert_id: required_int(fields, "last_insert_id")?,
    })
}

pub(crate) fn decode_snapshot_response(fields: &Json) -> Result<SnapshotResult, Error> {
    Ok(SnapshotResult {
        path: required_str(fields, "path")?.to_string(),
    })
}

pub(crate) fn decode_inspect_response(fields: &Json) -> Result<InspectResult, Error> {
    Ok(InspectResult {
        has_data: required_bool(fields, "has_data")?,
        version: required_optional_int(fields, "version")?,
        data_length: required_optional_int(fields, "data_length")?,
        source: required_str(fields, "source")?.to_string(),
        read_only: required_bool(fields, "read_only")?,
    })
}

pub(crate) fn decode_tables_response(fields: &Json) -> Result<TablesResult, Error> {
    Ok(TablesResult {
        tables: required_str_list(fields, "tables")?,
    })
}

pub(crate) fn decode_schema_response(fields: &Json) -> Result<SchemaResult, Error> {
    Ok(SchemaResult {
        schema: required_str_list(fields, "schema")?,
    })
}

pub(crate) fn decode_dump_response(fields: &Json) -> Result<DumpResult, Error> {
    Ok(DumpResult {
        sql: required_str(fields, "sql")?.to_string(),
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn hello_ok() {
        let h = decode_hello(br#"{"protocol":1,"version":"v0.1.1","product":"SanDBox"}"#).unwrap();
        assert_eq!(
            h,
            Hello {
                protocol: 1,
                version: "v0.1.1".into(),
                product: "SanDBox".into()
            }
        );
    }

    #[test]
    fn hello_does_not_validate_protocol_number() {
        // decode_hello only parses; comparing against PROTOCOL is the
        // caller's job (protocol.md, and Python's decode_hello docstring).
        let h = decode_hello(br#"{"protocol":999,"version":"x","product":"SanDBox"}"#).unwrap();
        assert_eq!(h.protocol, 999);
    }

    #[test]
    fn hello_missing_field_and_invalid_json() {
        assert!(decode_hello(br#"{"protocol":1,"version":"x"}"#).is_err());
        assert!(decode_hello(b"not json").is_err());
        assert!(decode_hello(b"[1,2,3]").is_err());
    }

    #[test]
    fn response_ok_true_carries_every_field() {
        let r =
            decode_response_line(br#"{"ok":true,"columns":["a"],"rows":[],"extra":"x"}"#).unwrap();
        assert!(r.ok);
        assert_eq!(r.fields.get("extra").and_then(Json::as_str), Some("x"));
    }

    #[test]
    fn response_ok_false_with_and_without_error_object() {
        let r = decode_response_line(
            br#"{"ok":false,"error":{"code":"bad_request","message":"nope"}}"#,
        )
        .unwrap();
        assert!(!r.ok);
        let e = response_error(&r.fields).unwrap();
        assert!(e.is_code("bad_request"));

        let r = decode_response_line(br#"{"ok":false}"#).unwrap();
        assert!(!r.ok);
        assert!(response_error(&r.fields).is_none());
    }

    #[test]
    fn blob_array_wrong_length_rejected() {
        assert!(decode_value(&Json::Array(vec![])).is_err());
        assert!(decode_value(&Json::Array(vec![Json::str("a"), Json::str("b")])).is_err());
    }

    #[test]
    fn bool_in_cell_rejected() {
        assert!(decode_value(&Json::Bool(true)).is_err());
    }

    #[test]
    fn rows_not_array_and_row_not_array_rejected_with_index() {
        assert!(decode_rows(&Json::Null).is_err());
        let err = decode_rows(&Json::Array(vec![Json::Null])).unwrap_err();
        assert!(matches!(err, Error::Protocol(m) if m.contains("rows[0]")));
    }

    #[test]
    fn decode_query_exec_snapshot() {
        let v = json::parse(br#"{"ok":true,"columns":["x"],"rows":[[1]]}"#).unwrap();
        let r = decode_query_response(&v).unwrap();
        assert_eq!(r.columns, vec!["x"]);
        assert_eq!(r.rows, vec![vec![Value::Integer(1)]]);

        let v = json::parse(br#"{"ok":true,"rows_affected":2,"last_insert_id":9}"#).unwrap();
        let r = decode_exec_response(&v).unwrap();
        assert_eq!(
            r,
            ExecResult {
                rows_affected: 2,
                last_insert_id: 9
            }
        );

        let v = json::parse(br#"{"ok":true,"path":"snap.bin"}"#).unwrap();
        assert_eq!(
            decode_snapshot_response(&v).unwrap(),
            SnapshotResult {
                path: "snap.bin".into()
            }
        );
    }

    #[test]
    fn decode_inspect_with_and_without_nulls() {
        let v = json::parse(
            br#"{"ok":true,"has_data":true,"version":3,"data_length":100,"source":"embedded","read_only":false}"#,
        )
        .unwrap();
        assert_eq!(
            decode_inspect_response(&v).unwrap(),
            InspectResult {
                has_data: true,
                version: Some(3),
                data_length: Some(100),
                source: "embedded".into(),
                read_only: false,
            }
        );

        let v = json::parse(
            br#"{"ok":true,"has_data":false,"version":null,"data_length":null,"source":"none","read_only":true}"#,
        )
        .unwrap();
        assert_eq!(
            decode_inspect_response(&v).unwrap(),
            InspectResult {
                has_data: false,
                version: None,
                data_length: None,
                source: "none".into(),
                read_only: true
            },
        );
    }

    #[test]
    fn decode_tables_schema_dump() {
        let v = json::parse(br#"{"ok":true,"tables":["a","b"]}"#).unwrap();
        assert_eq!(
            decode_tables_response(&v).unwrap(),
            TablesResult {
                tables: vec!["a".into(), "b".into()]
            }
        );

        let v = json::parse(br#"{"ok":true,"schema":["CREATE TABLE a(x)"]}"#).unwrap();
        assert_eq!(
            decode_schema_response(&v).unwrap(),
            SchemaResult {
                schema: vec!["CREATE TABLE a(x)".into()]
            }
        );

        let v = json::parse(br#"{"ok":true,"sql":"INSERT INTO a VALUES(1);"}"#).unwrap();
        assert_eq!(
            decode_dump_response(&v).unwrap(),
            DumpResult {
                sql: "INSERT INTO a VALUES(1);".into()
            }
        );
    }

    #[test]
    fn missing_required_field_is_protocol_error() {
        let v = json::parse(br#"{"ok":true,"columns":["x"]}"#).unwrap();
        assert!(decode_query_response(&v).is_err());
    }

    #[test]
    fn decode_value_large_integer_and_real() {
        let n = Json::Number(json::NumberToken::new("9223372036854775807".to_string()));
        assert_eq!(decode_value(&n).unwrap(), Value::Integer(i64::MAX));

        let n = Json::Number(json::NumberToken::new("88.0".to_string()));
        assert_eq!(decode_value(&n).unwrap(), Value::Real(88.0));

        let n = Json::Number(json::NumberToken::new("9e999".to_string()));
        assert_eq!(decode_value(&n).unwrap(), Value::Real(f64::INFINITY));
    }
}
